### PR TITLE
Handling Expect-100 Continue header

### PR DESCRIPTION
**Pull request by dhavalsingh**
…gins reading the body

### Description
Please describe your pull request. Thank you for contributing! You're the best.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [ ] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/client.rb [hunk 1]
```diff
@@ -342,12 +342,19 @@ def setup_body
       @body_read_start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :float_millisecond)
 
       if @env[HTTP_EXPECT] == CONTINUE
-        # TODO allow a hook here to check the headers before
-        # going forward
-        @io << HTTP_11_100
-        @io.flush
+        # Determine if a redirect is needed, this could be a function call that encapsulates
```
Comment (dhavalsingh) [inline lib/puma/client.rb:345]: Need to understand if this is what's expected

**Comment by dhavalsingh** [comment]
Will clean up the PR once I understand I am on the right path to solving this issue.
 @dentarg can you help as you created this issue?

**Comment by dentarg** [comment]
This doesn't look correct to me. We want to application to define the response (e.g. the 307 redirect).

**Comment by dentarg** [comment]
The best guidance I can give you is to look at how the Python `HTTPServer` does it (https://github.com/python/cpython/blob/3.12/Lib/http/server.py) as I have not researched this topic myself yet.

**Comment by dentarg** [comment]
Probably good to drive the implementation using tests.

**Comment by dhavalsingh** [comment]
> The best guidance I can give you is to look at how the Python `HTTPServer` does it (https://github.com/python/cpython/blob/3.12/Lib/http/server.py) as I have not researched this topic myself yet.

Ah, okay. Will look into this. Thanks a lot!

**Comment by dhavalsingh** [comment]
Where are client specific tests written? cant figure it out, or there arent any?

**Comment by dentarg** [comment]
[`test/test_web_server.rb`](https://github.com/puma/puma/blob/99ae85bbbe425d1949f9e8693993499aae02fce4/test/test_web_server.rb) is probably a good fit for this

File: test/test_web_server.rb [hunk 2]
```diff
@@ -99,6 +99,14 @@ def test_nonexistent_http_method
     socket.close
   end
 
+  def test_expect_100_header
```
Comment (dhavalsingh) [inline test/test_web_server.rb:102]: I should add some tests when response is non 200 also right?

**Comment by dhavalsingh** [comment]
Also, whats the diff between test_puma_server vs test_web_server tests?

File: lib/puma/client.rb [hunk 3]
```diff
@@ -342,10 +346,7 @@ def setup_body
       @body_read_start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :float_millisecond)
 
       if @env[HTTP_EXPECT] == CONTINUE
-        # TODO allow a hook here to check the headers before
-        # going forward
-        @io << HTTP_11_100
-        @io.flush
+        @http_expect_100_continue_header_present = true
```
Comment (dentarg) [inline lib/puma/client.rb:349]: Pretty sure we don't want to remove the current behaviour of responding quickly to `Expect: 100 Continue` and that allowing the app to define the response to be something the user opt-in to (via the Puma config probably)

Comment (nateberkopec) [reply lib/puma/client.rb:349]: Agreed!

Comment (nateberkopec) [reply lib/puma/client.rb:349]: (Or at least, we should keep that behavior UNLESS the user/app says they want something else).

**Comment by nateberkopec** [comment]
This seems like a fine place to start for me. Next step would be a config setting that allows the user to define what to do if a Expect: 100 is encountered.

**Comment by nateberkopec** [comment]
Closing as stale to allow others to pick this up or for OP to return

**Pull request closed without merging**